File: c implementation/code.c
#include <stdio.h>
#include <string.h>
#include <limits.h>
#define d 256
void search(char pattern[], char text[], int q) 
{
    int M = strlen(pattern);
    int N = strlen(text);
    int i, j;
    int p = 0;
    int t = 0;
    int h = 1;
    for (i = 0; i < M - 1; i++) 
	{
        h = (h * d) % q;
    }
    for (i = 0; i < M; i++) 
	{
        p = (d * p + pattern[i]) % q;
        t = (d * t + text[i]) % q;
    }
    for (i = 0; i <= N - M; i++) 
	{
        if (p == t) 
		{
            for (j = 0; j < M; j++) 
			{
                if (text[i + j] != pattern[j])
                    break;
            }
            if (j == M) 
			{
                printf("Pattern found at index %d\n", i);
            }
        }
        if (i < N - M) 
		{
            t = (d * (t - text[i] * h) + text[i + M]) % q;
            if (t < 0)
            {
                t = (t + q);
            }
        }
    }
}
int main() 
{
    FILE *inputFile;
    char txt[1000], pat[1000];
    int q;
    int num_test_cases, i;
    inputFile = fopen("input.txt", "r");
    if (inputFile == NULL) 
	{
        printf("Unable to open file.\n");
        return 1;
    }
    fscanf(inputFile, "%d", &num_test_cases);
    for (i = 0; i < num_test_cases; i++) 
	{
        fscanf(inputFile, "%s\n%s\n%d\n", txt, pat, &q);
        printf("Text: %s\n", txt);
        printf("Pattern: %s\n", pat);
        printf("Prime number q: %d\n", q);
        printf("\n");
        search(pat, txt, q);
        printf("\n");
    }
    fclose(inputFile);
    return 0;
}

File: python implementation/code.py
d = 256

def search(pattern, text, q):
    M = len(pattern)
    N = len(text)
    h = pow(d, M-1, q)
    p = 0
    t = 0
    for i in range(M):
        p = (d * p + ord(pattern[i])) % q
        t = (d * t + ord(text[i])) % q
    for i in range(N - M + 1):
        if p == t:
            for j in range(M):
                if text[i + j] != pattern[j]:
                    break
            else:
                print("Pattern found at index", i)
        if i < N - M:
            t = (d * (t - ord(text[i]) * h) + ord(text[i + M])) % q

def main():
    filename = "input.txt"
    try:
        with open(filename, 'r') as inputFile:
            num_test_cases = int(inputFile.readline().strip())
            print("Number of test cases:", num_test_cases)
            for _ in range(num_test_cases):
                line = inputFile.readline().split()
                if len(line) == 3:
                    txt, pat, q = line
                    print("Text:", txt)
                    print("Pattern:", pat)
                    print("Prime number q:", q)
                    print()
                    search(pat, txt, int(q))
                    print()
                else:
                    print("Invalid input format in file for test case.")
    except FileNotFoundError:
        print("Unable to open file.")

if __name__ == "__main__":
    main()
